test/e2e/route_advertisements.go

```diff
@@ -430,42 +412,81 @@ var _ = ginkgo.Describe("BGP: Pod to external server when CUDN Layer3 Network is
+				},
+				Spec: udnv1.ClusterUserDefinedNetworkSpec{
+					Network: udnv1.NetworkSpec{
+						Topology: "Layer3",
+						Layer3: &udnv1.Layer3Config{
+							Role: "Primary",
+							Subnets: generateL3Subnets(udnv1.Layer3Subnet{
+								CIDR:       "103.103.0.0/16",
+								HostSubnet: 24,
+							}, udnv1.Layer3Subnet{
+								CIDR:       "2014:100:200::0/60",
+								HostSubnet: 64,
+							}),
+						},
+					},
+				},
+			},
+			applycfgrav1.RouteAdvertisements("bgp-udn-layer3-network-ra").
+				WithSpec(
+					applycfgrav1.RouteAdvertisementsSpec().
+						WithAdvertisements(rav1.PodNetwork).
+						WithNetworkSelector(
+							v1.LabelSelector().WithMatchLabels(map[string]string{"bgp-udn-layer3-network": ""}),
+						),
+				),
+		),
+		ginkgo.Entry("layer2",
+			&udnv1.ClusterUserDefinedNetwork{
+				ObjectMeta: metav1.ObjectMeta{
+					GenerateName: "bgp-udn-layer2-network",
+					Labels:       map[string]string{"bgp-udn-layer2-network": ""},
+				},
+				Spec: udnv1.ClusterUserDefinedNetworkSpec{
+					Network: udnv1.NetworkSpec{
+						Topology: "Layer2",
+						Layer2: &udnv1.Layer2Config{
+							Role:    "Primary",
+							Subnets: generateL2Subnets("103.0.0.0/16", "2014:100::0/60"),
+						},
+					},
+				},
+			},
+			applycfgrav1.RouteAdvertisements("bgp-udn-layer2-network-ra").
+				WithSpec(
+					applycfgrav1.RouteAdvertisementsSpec().
+						WithAdvertisements(rav1.PodNetwork).
+						WithNetworkSelector(
+							v1.LabelSelector().WithMatchLabels(map[string]string{"bgp-udn-layer2-network": ""}),
+						),
+				),
+		),
+	)
 })
 
-func generateBGPCUDNManifest(testCudnName string, targetNamespaces ...string) string {
-	targetNs := strings.Join(targetNamespaces, ",")
-	return `
-apiVersion: k8s.ovn.org/v1
-kind: ClusterUserDefinedNetwork
-metadata:
-  name: ` + testCudnName + `
-  labels:
-    k8s.ovn.org/bgp-network: ""
-spec:
-  namespaceSelector:
-    matchExpressions:
-    - key: kubernetes.io/metadata.name
-      operator: In
-      values: [ ` + targetNs + ` ]
-  network:
-    topology: Layer3
-    layer3:
-      role: Primary
-      subnets: ` + generateCIDRforClusterUDN("103.103.0.0/16", "2014:100:200::0/60")
+func generateL3Subnets(v4, v6 udnv1.Layer3Subnet) []udnv1.Layer3Subnet {
+	var subnets []udnv1.Layer3Subnet
+	if isIPv4Supported() {
+		subnets = append(subnets, v4)
+	} else if isIPv6Supported() {
+		subnets = append(subnets, v6)
+	}
+	return subnets
 }
-
-func generateRAManifest(name string) string {
-	return `
-apiVersion: k8s.ovn.org/v1
-kind: RouteAdvertisements
-metadata:
-  name: ` + name + `
-spec:
-  networkSelector:
-    matchLabels:
-      k8s.ovn.org/bgp-network: ""
-  advertisements:
-    - "PodNetwork"`
+func generateL2Subnets(v4, v6 string) udnv1.DualStackCIDRs {
+	var subnets udnv1.DualStackCIDRs
+	if isIPv4Supported() {
+		subnets = append(subnets, udnv1.CIDR(v4))
+	} else if isIPv6Supported() {
+		subnets = append(subnets, udnv1.CIDR(v6))
+	}
+	return subnets
```

Comment:
This will always return a single stack ipv4 or ipv6 it should be something like 
```golang
var subnets udnv1.DualStackCIDRs
	if isIPv4Supported() {
		subnets = append(subnets, udnv1.CIDR(v4))
	} 

       if isIPv6Supported() {
		subnets = append(subnets, udnv1.CIDR(v6))
	}

	return subnets
```

Reply:
Thanks, fixed!